Sources/_MatchingEngine/Regex/DSLTree.swift

```diff
@@ -0,0 +1,332 @@
+  String, Range<String.Index>
+) -> (Any, Any.Type, String.Index)?
+
+// Character-set (post grapheme segmentation)
+typealias _CharacterPredicateInterface = (
+  (Character) -> Bool
+)
+
+/*
+
+ TODO: Use of syntactic types, like group kinds, is a
+ little suspect. We may want to figure out a model here.
+
+ TODO: Do capturing groups need explicit numbers?
+
+ TODO: Are storing closures better/worse than existentials?
+
+ */
+
+extension DSLTree.Node {
+  var children: [DSLTree.Node]? {
+    switch self {
+      
+    case let .alternation(v):   return v
+    case let .concatenation(v): return v
+
+    case let .convertedRegexLiteral(n, _):
+      // Treat this transparently
+      return n.children
+
+    case let .group(_, n):             return [n]
+    case let .groupTransform(_, n, _): return [n]
+    case let .quantification(_, _, n): return [n]
+
+    case let .conditional(_, t, f): return [t,f]
+
+    case .trivia, .empty, .stringLiteral, .regexLiteral,
+        .consumer, .consumerValidator, .characterPredicate,
+        .customCharacterClass, .unconvertedRegexLiteral,
+        .atom:
+      return []
+    }
+  }
+}
+
+extension DSLTree.Node {
+  var astNode: AST.Node? {
+    switch self {
+    case let .regexLiteral(ast):             return ast
+    case let .unconvertedRegexLiteral(ast):  return ast
+    case let .convertedRegexLiteral(_, ast): return ast
+    default: return nil
+    }
+  }
+}
+
+extension AST.Node {
+  /// Converts an AST node to a `convertedRegexLiteral` node.
+  var dslTreeNode: DSLTree.Node {
+    func wrap(_ node: DSLTree.Node) -> DSLTree.Node {
+      switch node {
+      case .convertedRegexLiteral:
+        assertionFailure("Double wrapping?")
+      default:
+        break
+      }
+      // TODO: Should we do this for the
+      // single-concatenation child too, or should?
+      // we wrap _that_?
+      return .convertedRegexLiteral(node, self)
+    }
+
+    // Convert the top-level node without wrapping
+    func convert() -> DSLTree.Node {
+      switch self {
+      case let .alternation(v):
+        let children = v.children.map(\.dslTreeNode)
+        return .alternation(children)
+
+      case let .concatenation(v):
+        // Coalesce adjacent children who can produce a
+        // string literal representation
+        let astChildren = v.children
+        func coalesce(
+          _ idx: inout Array<AST>.Index
+        ) -> String? {
+          var result = ""
+          while idx < astChildren.endIndex {
+            let atom: AST.Atom? = astChildren[idx].as()
+            guard let str = atom?.literalStringValue else {
+              break
+            }
+            result += str
+            astChildren.formIndex(after: &idx)
+          }
+          return result.isEmpty ? nil : result
+        }
+
+        // No need to nest single children concatenations
+        if astChildren.count == 1 {
+          return astChildren.first!.dslTreeNode
+        }
```

Comment:
When do single-child concatenations occur? I kind of wonder whether `AST.Concatenation` should enforce that `children.count >= 2` (and maybe have a utility that produces concat-or-self), or is it better for compatibility with the DSL to allow them?

Reply:
I’m in favor of enforcing a plural count.

Reply:
How would we enforce that in the DSL? Would we error out if given a single member? I worry there could be impact with sensible refactoring like quoting a literal run of characters vs listing each separately.

Reply:
The builder DSL always supports having a singular element in a block, e.g. `Regex { "a" }`. It's just that `buildBlock(_:)` (unary) never creates a concatenation in this scenario, but returns the argument instead.  As such, I'm not sure when else a singular concatenation would be created.